Better error message from jib:buildDocker when docker command requires sudo

Works fine when the `docker` command does not require sudo. However, the error message is not so helpful when it does.

```
$ docker
...
Run 'docker COMMAND --help' for more information on a command.

$ docker images
Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.37/images/json: dial unix /var/run/docker.sock: connect: permission denied

$ mvn -e compile jib:buildDocker
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 4.447 s
[INFO] Finished at: 2018-05-30T16:12:12-04:00
[INFO] Final Memory: 22M/214M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.1.8-SNAPSHOT:buildDocker (default-cli) on project test-artifact: Build to Docker daemon failed: Broken pipe -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.tools:jib-maven-plugin:0.1.8-SNAPSHOT:buildDocker (default-cli) on project test-artifact: Build to Docker daemon failed
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:213)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Build to Docker daemon failed
        at com.google.cloud.tools.jib.maven.BuildDockerMojo.execute(BuildDockerMojo.java:81)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        ... 20 more
Caused by: java.io.IOException: Broken pipe
        at java.io.FileOutputStream.writeBytes(Native Method)
        at java.io.FileOutputStream.write(FileOutputStream.java:326)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.write(BufferedOutputStream.java:126)
        at java.security.DigestOutputStream.write(DigestOutputStream.java:145)
        at com.google.cloud.tools.jib.hash.CountingDigestOutputStream.write(CountingDigestOutputStream.java:77)
        at org.apache.commons.compress.utils.CountingOutputStream.write(CountingOutputStream.java:48)
        at org.apache.commons.compress.utils.FixedLengthBlockOutputStream$BufferAtATimeOutputChannel.write(FixedLengthBlockOutputStream.java:241)
        at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.writeBlock(FixedLengthBlockOutputStream.java:92)
        at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.maybeFlush(FixedLengthBlockOutputStream.java:86)
        at org.apache.commons.compress.utils.FixedLengthBlockOutputStream.write(FixedLengthBlockOutputStream.java:122)
        at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.write(TarArchiveOutputStream.java:439)
        at com.google.common.io.ByteStreams.copy(ByteStreams.java:110)
        at com.google.cloud.tools.jib.tar.TarStreamBuilder.lambda$addEntry$0(TarStreamBuilder.java:71)
        at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeEntriesAsTarArchive(TarStreamBuilder.java:57)
        at com.google.cloud.tools.jib.blob.WriterBlob.writeTo(WriterBlob.java:36)
        at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:83)
        at com.google.cloud.tools.jib.builder.steps.BuildTarballAndLoadDockerStep.afterPushBaseImageLayerFuturesFuture(BuildTarballAndLoadDockerStep.java:118)
        at com.google.common.util.concurrent.CombinedFuture$CallableInterruptibleTask.runInterruptibly(CombinedFuture.java:181)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:57)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
        Suppressed: java.io.IOException: This archive contains unclosed entries.
                at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.finish(TarArchiveOutputStream.java:276)
                at org.apache.commons.compress.archivers.tar.TarArchiveOutputStream.close(TarArchiveOutputStream.java:293)
                at com.google.cloud.tools.jib.tar.TarStreamBuilder.$closeResource(TarStreamBuilder.java:60)
                at com.google.cloud.tools.jib.tar.TarStreamBuilder.writeEntriesAsTarArchive(TarStreamBuilder.java:60)
                ... 8 more
        Suppressed: java.io.IOException: Stream Closed
                at java.io.FileOutputStream.writeBytes(Native Method)
                at java.io.FileOutputStream.write(FileOutputStream.java:326)
                at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
                at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
                at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
                at com.google.cloud.tools.jib.docker.DockerClient.$closeResource(DockerClient.java:84)
                at com.google.cloud.tools.jib.docker.DockerClient.load(DockerClient.java:84)
                ... 6 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -75,13 +75,23 @@ public class DockerClient {
    *     href="https://docs.docker.com/engine/reference/commandline/load/">https://docs.docker.com/engine/reference/commandline/load/</a>
    * @return stdout from {@code docker}
    */
-  public String load(Blob imageTarballBlob) throws IOException, InterruptedException {
+  public String load(Blob imageTarballBlob) throws InterruptedException, IOException {
     // Runs 'docker load'.
     Process dockerProcess = docker("load");
 
     try (OutputStream stdin = dockerProcess.getOutputStream()) {
       imageTarballBlob.writeTo(stdin);
+
+    } catch (IOException ex) {
+      // This error usually happens when docker is not able to connect to the daemon socket.
+      try (InputStreamReader stderr =
+          new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
+        String error = CharStreams.toString(stderr);
+
+        throw new IOException("'docker load' command failed with error: " + error);
+      }
     }
+
     try (InputStreamReader stdout =
         new InputStreamReader(dockerProcess.getInputStream(), StandardCharsets.UTF_8)) {
       String output = CharStreams.toString(stdout);

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -89,7 +89,9 @@ public class DockerClient {
         try (InputStreamReader stderr =
             new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
           error = CharStreams.toString(stderr);
+
         } catch (IOException ignored) {
+          // This ignores exceptions from reading stderr and throws the original exception from writing to stdin.
           throw ex;
         }
         throw new IOException("'docker load' command failed with error: " + error, ex);

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -31,6 +31,20 @@ import java.util.function.Function;
 /** Calls out to the {@code docker} CLI. */
 public class DockerClient {
 
+  /** Accepts an {@link OutputStream}, possibly throwing an {@link IOException}. */
+  @FunctionalInterface
+  private interface OutputStreamConsumer {
+
+    void accept(OutputStream outputStream) throws IOException;
+  }
+
+  /** Handles an {@link IOException}, possibly throwing an {@link IOException}. */
+  @FunctionalInterface
+  private interface IOExceptionHandler {
+
+    void handle(IOException ex) throws IOException;
+  }
+
   /**
    * @param dockerSubCommand the subcommand to run after {@code docker}
    * @return the default {@link ProcessBuilder} factory for running a {@code docker} subcommand
@@ -40,6 +54,26 @@ public class DockerClient {
     dockerCommand.add("docker");
     dockerCommand.addAll(dockerSubCommand);
     return new ProcessBuilder(dockerCommand);
+  }
+
+  /**
+   * Writes a process's stdin.
+   *
+   * @param process the {@link Process} to
+   * @param stdinConsumer writes to a stdin {@link OutputStream}
+   * @param ioExceptionHandler handles any {@link IOException} returned by the {@code stdinConsumer}
+   */
+  private static void writeToStdin(
+      Process process, OutputStreamConsumer stdinConsumer, IOExceptionHandler ioExceptionHandler)
+      throws IOException {
+    try (OutputStream stdin = process.getOutputStream()) {
+      try {
+        stdinConsumer.accept(stdin);
+
+      } catch (IOException ex) {
+        ioExceptionHandler.handle(ex);
+      }
+    }
   }
 
   /** Factory for generating the {@link ProcessBuilder} for running {@code docker} commands. */
@@ -79,18 +113,18 @@ public class DockerClient {
     // Runs 'docker load'.
     Process dockerProcess = docker("load");
 
-    try (OutputStream stdin = dockerProcess.getOutputStream()) {
-      imageTarballBlob.writeTo(stdin);
-
-    } catch (IOException ex) {
-      // This error usually happens when docker is not able to connect to the daemon socket.
-      try (InputStreamReader stderr =
-          new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
-        String error = CharStreams.toString(stderr);
-
-        throw new IOException("'docker load' command failed with error: " + error);
-      }
-    }
+    writeToStdin(
+        dockerProcess,
+        imageTarballBlob::writeTo,
+        ioException -> {
+          // Tries to read from stderr.
+          try (InputStreamReader stderr =
+              new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
+            throw new IOException(
+                "'docker load' command failed with error: " + CharStreams.toString(stderr),
+                ioException);
+          }
+        });
 
     try (InputStreamReader stdout =
         new InputStreamReader(dockerProcess.getInputStream(), StandardCharsets.UTF_8)) {

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -31,20 +31,6 @@ import java.util.function.Function;
 /** Calls out to the {@code docker} CLI. */
 public class DockerClient {
 
-  /** Accepts an {@link OutputStream}, possibly throwing an {@link IOException}. */
-  @FunctionalInterface
-  private interface OutputStreamConsumer {
-
-    void accept(OutputStream outputStream) throws IOException;
-  }
-
-  /** Handles an {@link IOException}, possibly throwing an {@link IOException}. */
-  @FunctionalInterface
-  private interface IOExceptionHandler {
-
-    void handle(IOException ex) throws IOException;
-  }
-
   /**
    * @param dockerSubCommand the subcommand to run after {@code docker}
    * @return the default {@link ProcessBuilder} factory for running a {@code docker} subcommand
@@ -54,26 +40,6 @@ public class DockerClient {
     dockerCommand.add("docker");
     dockerCommand.addAll(dockerSubCommand);
     return new ProcessBuilder(dockerCommand);
-  }
-
-  /**
-   * Writes to a process's stdin.
-   *
-   * @param process the {@link Process} to
-   * @param stdinConsumer writes to a stdin {@link OutputStream}
-   * @param ioExceptionHandler handles any {@link IOException} returned by the {@code stdinConsumer}
-   */
-  private static void writeToStdin(
-      Process process, OutputStreamConsumer stdinConsumer, IOExceptionHandler ioExceptionHandler)
-      throws IOException {
-    try (OutputStream stdin = process.getOutputStream()) {
-      try {
-        stdinConsumer.accept(stdin);
-
-      } catch (IOException ex) {
-        ioExceptionHandler.handle(ex);
-      }
-    }
   }
 
   /** Factory for generating the {@link ProcessBuilder} for running {@code docker} commands. */
@@ -113,18 +79,22 @@ public class DockerClient {
     // Runs 'docker load'.
     Process dockerProcess = docker("load");
 
-    writeToStdin(
-        dockerProcess,
-        imageTarballBlob::writeTo,
-        ioException -> {
-          // Tries to read from stderr.
-          try (InputStreamReader stderr =
-              new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
-            throw new IOException(
-                "'docker load' command failed with error: " + CharStreams.toString(stderr),
-                ioException);
-          }
-        });
+    try (OutputStream stdin = dockerProcess.getOutputStream()) {
+      try {
+        imageTarballBlob.writeTo(stdin);
+
+      } catch (IOException ex) {
+        // Tries to read from stderr.
+        String error;
+        try (InputStreamReader stderr =
+            new InputStreamReader(dockerProcess.getErrorStream(), StandardCharsets.UTF_8)) {
+          error = CharStreams.toString(stderr);
+        } catch (IOException ignored) {
+          throw ex;
+        }
+        throw new IOException("'docker load' command failed with error: " + error, ex);
+      }
+    }
 
     try (InputStreamReader stdout =
         new InputStreamReader(dockerProcess.getInputStream(), StandardCharsets.UTF_8)) {

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -57,7 +57,7 @@ public class DockerClient {
   }
 
   /**
-   * Writes a process's stdin.
+   * Writes to a process's stdin.
    *
    * @param process the {@link Process} to
    * @param stdinConsumer writes to a stdin {@link OutputStream}

--- a/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
+++ b/jib-core/src/main/java/com/google/cloud/tools/jib/docker/DockerClient.java
@@ -91,7 +91,8 @@ public class DockerClient {
           error = CharStreams.toString(stderr);
 
         } catch (IOException ignored) {
-          // This ignores exceptions from reading stderr and throws the original exception from writing to stdin.
+          // This ignores exceptions from reading stderr and throws the original exception from
+          // writing to stdin.
           throw ex;
         }
         throw new IOException("'docker load' command failed with error: " + error, ex);

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
@@ -21,6 +21,7 @@ import com.google.common.io.ByteStreams;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Paths;
 import org.junit.Assert;
@@ -85,7 +86,32 @@ public class DockerClientTest {
   }
 
   @Test
-  public void testLoad_fail() throws InterruptedException {
+  public void testLoad_stdinFail() throws InterruptedException {
+    DockerClient testDockerClient = new DockerClient(ignored -> mockProcessBuilder);
+
+    Mockito.when(mockProcess.getOutputStream())
+        .thenReturn(
+            new OutputStream() {
+
+              @Override
+              public void write(int b) throws IOException {
+                throw new IOException();
+              }
+            });
+    Mockito.when(mockProcess.getErrorStream())
+        .thenReturn(new ByteArrayInputStream("error".getBytes(StandardCharsets.UTF_8)));
+
+    try {
+      testDockerClient.load(Blobs.from("jib"));
+      Assert.fail("Write should have failed");
+
+    } catch (IOException ex) {
+      Assert.assertEquals("'docker load' command failed with error: error", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testLoad_stdoutFail() throws InterruptedException {
     DockerClient testDockerClient = new DockerClient(ignored -> mockProcessBuilder);
     Mockito.when(mockProcess.waitFor()).thenReturn(1);
 

--- a/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
+++ b/jib-core/src/test/java/com/google/cloud/tools/jib/docker/DockerClientTest.java
@@ -21,6 +21,7 @@ import com.google.common.io.ByteStreams;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Paths;
@@ -107,6 +108,39 @@ public class DockerClientTest {
 
     } catch (IOException ex) {
       Assert.assertEquals("'docker load' command failed with error: error", ex.getMessage());
+    }
+  }
+
+  @Test
+  public void testLoad_stdinFail_stderrFail() throws InterruptedException {
+    DockerClient testDockerClient = new DockerClient(ignored -> mockProcessBuilder);
+    IOException expectedIOException = new IOException();
+
+    Mockito.when(mockProcess.getOutputStream())
+        .thenReturn(
+            new OutputStream() {
+
+              @Override
+              public void write(int b) throws IOException {
+                throw expectedIOException;
+              }
+            });
+    Mockito.when(mockProcess.getErrorStream())
+        .thenReturn(
+            new InputStream() {
+
+              @Override
+              public int read() throws IOException {
+                throw new IOException();
+              }
+            });
+
+    try {
+      testDockerClient.load(Blobs.from("jib"));
+      Assert.fail("Write should have failed");
+
+    } catch (IOException ex) {
+      Assert.assertTrue(expectedIOException == ex);
     }
   }
 

